### PR TITLE
[uefi] add mouse pointer supported.

### PR DESCRIPTION
**Pull request by sky5454**
[example/uefi-demo] add mouse pointer supported. now you can see the mouse which is a rectangle move as you want. and you can click it then the slint element would be effected.

### CODE REVIEW

File: examples/uefi-demo/main.rs [hunk 1]
```diff
@@ -213,21 +248,83 @@ impl slint::platform::Platform for Platform {
...
+
+        let ptr = POINTER.load(Ordering::Relaxed);
+        let mpointer = unsafe { &mut *ptr };
+        let conpointer = unsafe { &*ptr };
+        let mouse_mode = conpointer.mode();
+
         loop {
             slint::platform::update_timers_and_animations();
 
-            while let Some(key) = get_key_press() {
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:224]: the previous code had a `while` there.  Why did you remove it. It allowed to process several key per frame if the user presses keys faster than we are able to render frames.

**Review by ogoffart — COMMENTED** [review]
Thanks a lot for the patch!

Would be nice to use an actual mouse pointer. eg this image used in the linuxkms. 

https://github.com/slint-ui/slint/blob/master/internal/backends/linuxkms/mouse-pointer.svg


But I think the API needed to have the mouse pointer painted by the renderer like the linuxkms is not part of the API.

 @tronical : how could we expose this API?

**Comment by sky5454** [comment]
> Thanks a lot for the patch!
> 
> Would be nice to use an actual mouse pointer. eg this image used in the linuxkms.
> 
> https://github.com/slint-ui/slint/blob/master/internal/backends/linuxkms/mouse-pointer.svg
> 
> But I think the API needed to have the mouse pointer painted by the renderer like the linuxkms is not part of the API.
> 
> @tronical : how could we expose this API?

yes, but I can't still found a render for UEFI which is no_std rust. 
And that't  the next step I would do it. the key is the we should found the no_std render.

File: examples/uefi-demo/main.rs (hunk 1 above)
Comment (sky5454) [reply examples/uefi-demo/main.rs:224]: > the previous code had a there. Why did you remove it. It allowed to process several key per frame if the user presses keys faster than we are able to render frames.`while`

thanks, now I know the code. I would wrap input with a loop later while next pr edit.

**Comment by sky5454** [comment]
I got the cursor now! thanks for the lib minipng. Review the code and recive it please. thank you!
run_at_uefi

**Comment by tronical** [comment]
> Thanks a lot for the patch!
> 
> Would be nice to use an actual mouse pointer. eg this image used in the linuxkms.
> 
> https://github.com/slint-ui/slint/blob/master/internal/backends/linuxkms/mouse-pointer.svg
> 
> But I think the API needed to have the mouse pointer painted by the renderer like the linuxkms is not part of the API.
> 
> @tronical : how could we expose this API?

I think synthetic mouse cursor rendering should be implemented cross-platform/cross-renderer as opt-in feature.

Even MCUs come with USB ports nowadays and sometimes users attach mice and keyboards.

You had a very good idea in the software renderer PR:  make the mouse cursor an item tree with an Image that’s rendered after the other item trees, so minimize renderer impact and integrate with partial rendering.

What if we had a Boolean property (synthetic-mouse-cursor) somewhere in Window or the backend?

If true, the cross-platform code in core lib could maintain a synthetic Image item tree above the scene.

Backends need to be able to toggle this property (set it to false again when unplugging a mouse).

File: examples/uefi-demo/main.rs [hunk 2]
```diff
@@ -61,6 +79,20 @@ fn timer_freq() -> u64 {
...
+fn pointer_init() {
+    // mouse pointer
+    let bs = st().boot_services();
+    let handle = bs.get_handle_for_protocol::<Pointer>().expect("miss Pointer protocol");
+
+    let mut pointer =
+        bs.open_protocol_exclusive::<Pointer>(handle).expect("can't open Pointer protocol.");
+
+    pointer.reset(false).expect("Failed to reset pointer device.");
+    info!("pointer inited, mode = {:?}.", pointer.mode());
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:91]: do we want to keep these calls to `info!`?

File: examples/uefi-demo/main.rs [hunk 3]
```diff
@@ -159,6 +198,57 @@ impl software_renderer::TargetPixel for SlintBltPixel {
...
+
+    fn blend_blt_pixel(&self, background: &mut BltPixel) {
+        // Alpha Blending
+        // Result=Foreground×α+Background×(1−α)
+        let alpha = self.0[3] as f32 / 255.0;
+        let r = self.0[0] as f32;
+        let g = self.0[1] as f32;
+        let b = self.0[2] as f32;
+
+        // 根据透明度对叠加后的颜色进行计算
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:236]: could we keep the comments in english?

File: examples/uefi-demo/main.rs [hunk 4]
```diff
@@ -159,6 +198,57 @@ impl software_renderer::TargetPixel for SlintBltPixel {
...
+        let blended_r = ((1.0 - alpha) * background.red as f32 + alpha * r) as u8;
+        let blended_g = ((1.0 - alpha) * background.green as f32 + alpha * g) as u8;
+        let blended_b = ((1.0 - alpha) * background.blue as f32 + alpha * b) as u8;
+
+        background.red = blended_r;
+        background.green = blended_g;
+        background.blue = blended_b;
+        // BltPixel::new(blended_r, blended_g, blended_b)
+    }
+    // #[allow(dead_code)]
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:246]: this can be removed

File: examples/uefi-demo/main.rs [hunk 5]
```diff
@@ -159,6 +198,57 @@ impl software_renderer::TargetPixel for SlintBltPixel {
...
+        PngRGBAPixel([254, 254, 254, 0])
+    }
+    #[allow(dead_code)]
+    fn from_array(&mut self, pixel_data: [u8; 4]) {
+        self.0 = pixel_data;
+    }
+    fn from_rgba(&mut self, r: u8, g: u8, b: u8, a: u8) {
+        self.0 = [r, g, b, a];
+    }
+    #[allow(dead_code)]
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:217]: unused functions can be removed

File: examples/uefi-demo/main.rs [hunk 6]
```diff
@@ -11,14 +11,32 @@ use alloc::{
...
 use uefi_services::system_table;
 
+use slint::{platform::software_renderer, SharedString};
+
 slint::include_modules!();
 
+static POINTER: AtomicPtr<ScopedProtocol<'static, Pointer>> = AtomicPtr::new(core::ptr::null_mut());
+
+/* framebuffer coordinate bound limit for uefi-rs GOP */
+static BOUND_SPACE: usize = 0;
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:38]: should probably be `const` instead of static

File: examples/uefi-demo/main.rs [hunk 7]
```diff
@@ -11,14 +11,32 @@ use alloc::{
...
-use core::{slice, time::Duration};
-use slint::{platform::software_renderer, SharedString};
-use uefi::{prelude::*, proto::console::gop::BltPixel, Char16};
 use uefi_services::system_table;
 
+use slint::{platform::software_renderer, SharedString};
+
 slint::include_modules!();
 
+static POINTER: AtomicPtr<ScopedProtocol<'static, Pointer>> = AtomicPtr::new(core::ptr::null_mut());
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:35]: this could probably be named MOUSE_POINTER because i thought this was because it was a memory pointer (AtomicPtr) but that was misleading.

File: examples/uefi-demo/main.rs [hunk 8]
```diff
@@ -61,6 +79,20 @@ fn timer_freq() -> u64 {
...
+    // mouse pointer
+    let bs = st().boot_services();
+    let handle = bs.get_handle_for_protocol::<Pointer>().expect("miss Pointer protocol");
+
+    let mut pointer =
+        bs.open_protocol_exclusive::<Pointer>(handle).expect("can't open Pointer protocol.");
+
+    pointer.reset(false).expect("Failed to reset pointer device.");
+    info!("pointer inited, mode = {:?}.", pointer.mode());
+    let raw_ptr = Box::into_raw(Box::new(pointer));
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:92]: why do we make it a raw pointer instead of just storing the Box?

File: examples/uefi-demo/main.rs [hunk 9]
```diff
@@ -133,8 +165,15 @@ fn wait_for_input(max_timeout: Option<Duration>) {
     {
         // SAFETY: The cloned handles are only used to wait for further input events and
         // are then immediately dropped.
-        let mut events =
-            unsafe { [st().stdin().wait_for_key_event().unsafe_clone(), timer.unsafe_clone()] };
+        let ptr = POINTER.load(Ordering::Relaxed);
+        let pointer_ref = unsafe { &*ptr };
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:169]: we should remove this unsafe here if we store the Box instead of a raw pointer.

File: examples/uefi-demo/main.rs [hunk 10]
```diff
@@ -229,6 +352,52 @@ impl slint::platform::Platform for Platform {
...
+                });
+                self.window.dispatch_event(slint::platform::WindowEvent::PointerReleased {
+                    position,
+                    button,
+                });
+                is_mouse_move = true;
+            }
+
+            if is_mouse_move {
+                self.window.request_redraw();
```
Comment (ogoffart) [inline examples/uefi-demo/main.rs:398]: this doesn't mark the region dirty. I think that is missing.

**Review by ogoffart — COMMENTED** [review]
> I think synthetic mouse cursor rendering should be implemented cross-platform/cross-renderer as opt-in feature.

Right. But this is a bit too much for this PR though. I think we could just merge it with the current trick.

Although there is still a bit of cleanup to do.

I added a few comment inline.

Also, these image (the images into docs/), and especially the large .gif file, doesn't belong in the git repository. The git history need to be cleaned to not contain them. They can be hosted on github like the other screenshot.  (just drag them in a comment to get an URL)

**Comment by tronical** [comment]
> Right. But this is a bit too much for this PR though. I think we could just merge it with the current trick.

Agreed.

File: examples/uefi-demo/main.rs (hunk 2 above)
Comment (sky5454) [reply examples/uefi-demo/main.rs:91]: > do we want to keep these calls to `info!`?

yes, OVMF implements UEFI pointer protocol but it can't move the pointer. by serial log could see the different `pointer.mode()` print. so keep those code please.

File: examples/uefi-demo/main.rs (hunk 10 above)
Comment (sky5454) [reply examples/uefi-demo/main.rs:398]: when mouse pointer is moving, the slint-ui doesn't feel it. so I must request redraw. or the mouse would move everywhere makes screen dirty while slint-ui no refresh it.
So I suggest that slint-ui should provide the mouse pointer. The better way which may as [tronical](https://github.com/tronical)  say https://github.com/slint-ui/slint/pull/5034#issuecomment-2054159637

**Comment by sky5454** [comment]
Perhaps Slint could provides an RGBA image overlay, which allows developers to add custom pixels to the framebuffer, parameters could be included: (the starting and ending coordinates of the rectangle, the pixel structure of the image, and boolean whether to rerender)

That is not only use in cursor/mouse/mice pointer, but also lots of application scenarios, which give developer ability to call the native driver layer function.

File: examples/uefi-demo/main.rs (hunk 9 above)
Comment (sky5454) [reply examples/uefi-demo/main.rs:169]: > we should remove this unsafe here if we store the Box instead of a raw pointer.
> why do we make it a raw pointer instead of just storing the Box?

help me to fix it please

Comment (ogoffart) [reply examples/uefi-demo/main.rs:169]: This just doesn't look safe.

Could you perhaps call get_handle_for_protocol and get_handle_for_protocol everytime, or are these expensive calls?

**Comment by ogoffart** [comment]
Could you please remove the images from docs/ that you added in this PR?
